### PR TITLE
Docker: Major refactoring

### PR DESCRIPTION
**Pull request by oschuett**

### CODE REVIEW

File: tools/docker/scripts/ci_entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+          --exclude=/tools/toolchain/         \
+          --executability                     \
+          --ignore-times                      \
+          --update                            \
+          --verbose                           \
+          --recursive                         \
+          --checksum                          \
+          /mnt/cp2k/  /workspace/cp2k/
+else
+    echo "Neighter GIT_REF nor /mnt/cp2k found - aborting."
```
Comment (dev-zero) [inline tools/docker/scripts/ci_entrypoint.sh]: `neighter` -> `neither`

File: tools/docker/scripts/ci_entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    git checkout "${GIT_REF}"
+    git submodule update --init --recursive
+    git --no-pager log -1 --pretty='%nCommitSHA: %H%nCommitTime: %ci%nCommitAuthor: %an%nCommitSubject: %s%n' | tee -a $REPORT
+
+elif [ -d  /mnt/cp2k ]; then
+    echo -e "\n========== Copying Changed Files =========="
+    #TODO: Exclusion rules could also come from .gitigore
+    rsync --exclude="*~"                      \
+          --exclude=".*/"                     \
+          --exclude="*.pyc"                   \
```
Comment (dev-zero) [inline tools/docker/scripts/ci_entrypoint.sh]: `--exclude="*.py[cod]" --exclude="__pycache__"`

File: tools/docker/scripts/ci_entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    cd /workspace/cp2k
+    git fetch origin "${GIT_BRANCH}"
+    git checkout "${GIT_REF}"
+    git submodule update --init --recursive
+    git --no-pager log -1 --pretty='%nCommitSHA: %H%nCommitTime: %ci%nCommitAuthor: %an%nCommitSubject: %s%n' | tee -a $REPORT
+
+elif [ -d  /mnt/cp2k ]; then
+    echo -e "\n========== Copying Changed Files =========="
+    #TODO: Exclusion rules could also come from .gitigore
+    rsync --exclude="*~"                      \
```
Comment (dev-zero) [inline tools/docker/scripts/ci_entrypoint.sh]: you might want to either `rm -rf /workspace/cp2k/` or use `--delete` to make sure to clear removed files, especially with a build system which picks up files automatically

File: tools/docker/scripts/ci_entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+          --exclude="*.pyc"       \
+          --exclude="/build"      \
+          --exclude="/install"    \
+          --executability         \
+          --ignore-times          \
+          --update                \
+          --verbose               \
+          --recursive             \
+          --checksum              \
+          /workspace/cp2k/tools/toolchain/  /opt/cp2k-toolchain/
```
Comment (dev-zero) [inline tools/docker/scripts/ci_entrypoint.sh]: dito. You might want to factor out common rsync arguments or the rsync call itself

File: tools/docker/scripts/ci_entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+fi
+
+echo -e "\n========== Running Test =========="
+cd /workspace
+"$@" 2>&1 | tee -a $REPORT &  # Launch in the background.
+TEST_PID=$!
+trap '[ -z $TEST_PID] || kill $TEST_PID' SIGHUP SIGINT SIGQUIT SIGTERM
+
+# Upload preliminary report every 30s while test is running.
+while [ -e /proc/${TEST_PID} ]; do
```
Comment (dev-zero) [inline tools/docker/scripts/ci_entrypoint.sh]: why not use the `jobs` builtin instead? Would be more portable.

File: tools/docker/scripts/test_regtest.sh [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash -e
+
+# author: Ole Schuett
+
+ARCH=$1
+VERSION=$2
+
```
Comment (dev-zero) [inline tools/docker/scripts/test_regtest.sh:12]: if you are expecting exactly two arguments, assert for that:

```bash
[ $# -eq 2 ] || (echo "Usage: ..." && exit 1)
```

File: tools/docker/scripts/test_python.sh [hunk 7]
```diff
@@ -1,62 +1,54 @@
...
 # find executable python scripts
-ALL_SCRIPTS=`find ./src/ ./tools/ -name "*.py"  -executable`
-ESSENTIAL_SCRIPTS=`find ./tools/build_utils -name "*.py"  -executable`
+ALL_SCRIPTS=$(find ./src/ ./tools/ -name "*.py"  -executable)
+ESSENTIAL_SCRIPTS=$(find ./tools/build_utils -name "*.py"  -executable)
 ESSENTIAL_SCRIPTS="$ESSENTIAL_SCRIPTS ./src/dbcsr/libsmm_acc/libcusmm/generate.py"
 
 # python 2.6
-run_selftests python2.6 $ESSENTIAL_SCRIPTS
+run_selftests python2.6 "${ESSENTIAL_SCRIPTS}"
```
Comment (dev-zero) [inline tools/docker/scripts/test_python.sh:42]: Please drop 2.6 testing: Python 2.6 is EOL and popular projects have even started dropping Python 2.7 support

File: tools/docker/scripts/test_coverage.sh [hunk 8]
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash -e
+
+# author: Ole Schuett
+
+ARCH=local_coverage
+VERSION=$1
+
```
Comment (dev-zero) [inline tools/docker/scripts/test_coverage.sh:12]: please check number of arguments passed, see comment above

File: tools/docker/scripts/ci_entrypoint.sh (hunk 3 above)
Comment (oschuett) [reply tools/docker/scripts/ci_entrypoint.sh]: Good catch - thanks!

File: tools/docker/scripts/test_python.sh (hunk 7 above)
Comment (oschuett) [reply tools/docker/scripts/test_python.sh:42]: It probably comes at no surprise to you that I'm reluctant to drop Python 2.6 support ;-)

However, we should really find a better way to settle these kind of arguments. So, I propose we simply ask our users (#90).

File: tools/docker/scripts/ci_entrypoint.sh (hunk 5 above)
Comment (oschuett) [reply tools/docker/scripts/ci_entrypoint.sh]: The best I could come up with is `while jobs -p | grep "${TEST_PID}" > /dev/null; do`, which I find less readable. I don't think portability matters that much as we're using Docker.

Comment (dev-zero) [reply tools/docker/scripts/ci_entrypoint.sh]: AFAIK Docker on Mac needs special flags to get a `/proc` in a container

Comment (dev-zero) [reply tools/docker/scripts/ci_entrypoint.sh]: plus you might be checking any PID (after PID reuse) and not necessarily the child PID, while `jobs` should give the PID only if it was spawned from this shell.

Comment (oschuett) [reply tools/docker/scripts/ci_entrypoint.sh]: You're absolutely right. I was actually checking the PID of the `tee` process. I just pushed a new and improved version.

**Comment by dev-zero** [comment]
Unless you plan to squash your commits, please read https://github.com/cp2k/cp2k/wiki/Contribution-Guidelines#commit-messages

**Comment by oschuett** [comment]
Don't worry. I'll squash them. Sorry for the noise.

**Pull request merged**